docs/internal.rst

```diff
@@ -6,6 +6,68 @@ Intermediate representation
 Use ``ti.init(print_ir=True)`` to print IR on the console.
 
 
+Data Structure Node Organization
+--------------------------------
+
+The internal organization of data structure nodes can be confusing.
+It is important to distinguish the concept of **containers**, **cells** and **components**.
+
+- A **container** can have multiple **cells**. The numbers of **cells** are recommended to be powers of two.
+- A **cell** can have multiple **components**.
+- Each **component** is a **container** of a lower-level SNode.
+
+Consider the following example:
+
+.. code-block:: python
+
+    # misc/listgen_demo.py
+
+    x = ti.var(ti.i32)
+    y = ti.var(ti.i32)
+    z = ti.var(ti.i32)
+
+    S0 = ti.root
+    S1 = S0.pointer(ti.i, 4)
+
+    S2 = S1.dense(ti.i, 8)
+    S2.place(x, y) # S3: y; S4: z
+
+    S5 = S1.dense(ti.i, 8)
+    S5.place(z) # S6: z
+
+
+- The whole data structure is a ``S0root`` **container**, containing
+
+  - ``1x`` ``S0root`` **cell**, each has only one **component**, which is
+
+    - A ``S1pointer`` **container**, containing
+
+      - 4x ``S1pointer`` **cells**, each with two **components**, which are
+
+        - A ``S2dense`` **container**, containing
+
+          - 8x ``S2dense`` **cells**, each with two **components**, which are
+
+            - ``x: ti.i32``
+            - ``y: ti.i32``
+
+        - A ``S5dense`` **container**, containing
+
+          - 8x ``S5dense`` **cells**, each with one **component**, which is
+
+            - ``z: ti.i32``
+
```

Comment:
Personally I find it also useful to know the total number of each cells from root:

* For `S1`, this should be 1x4 = 4 cells
* For `S2`, this should be 1x4x8 = 32 cells

This is very obvious, but still, there are places where we need to distinguish between #cells / component and total #cells...